mlflow/utils/environment.py

```diff
@@ -453,18 +454,27 @@ def _is_mlflow_requirement(requirement_string):
             )
 
 
-def _generate_mlflow_version_pinning():
+def _generate_mlflow_version_pinning() -> str:
     """
     Determines the current MLflow version that is installed and adds a pinned boundary version range
-    for mlflow. The upper bound is a cap on the next major revision. The lower bound is a cap on
-    the current installed minor version(i.e., 'mlflow<3,>=2.1')
-    :return: string for MLflow dependency version
+    for mlflow.
+
+    :return: string for MLflow dependency version.
     """
+    if _MLFLOW_TESTING.get():
+        # The local PyPI server should be running. It serves a wheel for the current MLflow version.
+        return f"mlflow=={VERSION}"
+
     version = Version(VERSION)
-    # The version on master is always a micro-version ahead of the latest release and can't be
-    # installed from PyPI. We therefore subtract 1 from the micro version when running tests.
-    offset = -1 if version.is_devrelease else 0
-    return f"mlflow=={version.major}.{version.minor}.{version.micro + offset}"
+    if not version.is_devrelease:
+        # mlflow is installed from PyPI.
+        return f"mlflow=={VERSION}"
+
+    # mlflow installed from the source for development purposes. A dev version (e.g., 2.8.1.dev0)
+    # is always a micro-version ahead of the latest release (unless it's manually modified)
+    # and can't be installed from PyPI. We therefore subtract 1 from the micro version when running
+    # tests.
```

Comment:
We always have _MLFLOW_TESTING set during testing, what kind of scenario should we expect to reach here?

Reply:
Good question. For example, when we install mlflow on Databricks.

Reply:
Ah I see, thx for the explanation! Could you also update the comment accordingly?

Reply:
Updated!